File: 1584-min-cost-to-connect-all-points/1584-min-cost-to-connect-all-points.cpp
class Solution {
public:
    int minCostConnectPoints(vector<vector<int>>& arr) {
        int n=arr.size();
        
        int c=0,res=0;
        priority_queue<pair<int,int>>pq;
        int i=0;
        
        vector<int>vis(n,0);
        
        while(++c<n){
            vis[i]=1;
            for(int j=0;j<n;j++){
                if(vis[j]==0)
                    pq.push({-(abs(arr[i][0]-arr[j][0])+abs(arr[i][1]-arr[j][1])),j});
            }
            while(vis[pq.top().second])
                pq.pop();
            
            res-=pq.top().first;
            i=pq.top().second;
            pq.pop();
        }
        
        return res;
    }
};

File: 1584-min-cost-to-connect-all-points/1584-min-cost-to-connect-all-points.java
class Solution {
    
    int find(int []arr, int i){
        if(arr[i]==i)
            return i;
        return find(arr,arr[i]);
    }
    
    void union(int arr[],int i,int j){
        i=find(arr,i);
        j=find(arr,j);
        
        arr[i]=j;
    }
    
    public int minCostConnectPoints(int[][] points) {
        int len=points.length;
        
        ArrayList<Integer[]>arr=new ArrayList<>();
        
        for(int i=0;i<len;i++){
            for(int j=i+1;j<len;j++){
                arr.add(new Integer[]{i,j,Math.abs(points[i][0]-points[j][0])+
                                          Math.abs(points[i][1]-points[j][1])});
            }
        }
        
        Collections.sort(arr,(a,b)->{return a[2]-b[2];});
        
        int brr[]=new int[len];
        
        for(int i=0;i<len;i++){
            brr[i]=i;
        }
        
        int res=0;
        
        for(int i=0;i<arr.size();i++){
            int a=arr.get(i)[0];
            int b=arr.get(i)[1];
            int c=arr.get(i)[2];
            
            if(find(brr,a)!=find(brr,b)){
                union(brr,a,b);
                res+=c;
            }
        }
        
        return res;
    }
}